### PR TITLE
fix: Reload the app when there is a javascript error and a new version of the app

### PR DESCRIPTION
**Pull request by leggechr**
The initial implementation by @moodysalem was done in: https://github.com/Uniswap/interface/pull/3435. The problem there was if there was no update from the app it would get stuck refreshing infinitely.

I copied that PR but added a check to ensure that there is actually a new version being returned from the update and only in that case will the app update. I also had to add a few more lines that appear to be required to make the refresh happen properly which I copied from: https://github.com/facebook/create-react-app/issues/5316#issuecomment-591075209

### CODE REVIEW

**Review by NoahZinsmeister — APPROVED** [review]
this looks good! i know you said you were able to test this locally - how hard do you think it would be to add a unit test? or do you think it makes more sense to try to land tests once the fully featured service worker flow lands?

**Comment by leggechr** [comment]
@NoahZinsmeister the local testing I did involved building and serving the binary and then using the devtools to force an update so it doesn't translate easily to unit tests. I'd prefer to wait until we decide what to do with the rest of the service worker and then I can implement unit tests for the whole thing.

File: src/components/ErrorBoundary/index.tsx [hunk 1]
```diff
@@ -49,13 +49,39 @@ type ErrorBoundaryState = {
 
 const IS_UNISWAP = window.location.hostname === 'app.uniswap.org'
 
+async function updateServiceWorker(): Promise<ServiceWorkerRegistration> {
+  const ready = await navigator.serviceWorker.ready
+  // the return type of update is incorrectly indicated as Promise<void>
```
Comment (zzmp) [inline src/components/ErrorBoundary/index.tsx:54]: nit: "incorrectly typed as"

File: src/components/ErrorBoundary/index.tsx [hunk 2]
```diff
@@ -49,13 +49,39 @@ type ErrorBoundaryState = {
...
+  // https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/update
```
Comment (zzmp) [inline src/components/ErrorBoundary/index.tsx:55]: nit: "See <url>"

File: src/components/ErrorBoundary/index.tsx [hunk 3]
```diff
@@ -49,13 +49,39 @@ type ErrorBoundaryState = {
...
 export default class ErrorBoundary extends React.Component<unknown, ErrorBoundaryState> {
   constructor(props: unknown) {
     super(props)
     this.state = { error: null }
   }
 
   static getDerivedStateFromError(error: Error): ErrorBoundaryState {
+    updateServiceWorker()
+      .then(async (registration) => {
+        // We want to refresh only if we detect a new service worker is
```
Comment (zzmp) [inline src/components/ErrorBoundary/index.tsx:68]: Our [printWidth is set to 120](https://github.com/Uniswap/interface/blob/main/.prettierrc#L4), so this comment can be a single line.

File: src/components/ErrorBoundary/index.tsx [hunk 4]
```diff
@@ -49,13 +49,39 @@ type ErrorBoundaryState = {
...
+        // waiting to be activated.
+        // Details about it: https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle
```
Comment (zzmp) [inline src/components/ErrorBoundary/index.tsx:70]: nit: "See <url>"

File: src/components/ErrorBoundary/index.tsx [hunk 5]
```diff
@@ -49,13 +49,39 @@ type ErrorBoundaryState = {
...
+        if (registration?.waiting) {
+          await registration.unregister()
+
+          // Makes Workbox call skipWaiting()
```
Comment (zzmp) [inline src/components/ErrorBoundary/index.tsx:74]: Can you link to documentation in this comment? This isn't a typed API, so there's no real inline documentation.

File: src/components/ErrorBoundary/index.tsx (hunk 3 above)
Comment (zzmp) [inline src/components/ErrorBoundary/index.tsx:68]: Can you update this comment to include the rationale behind using skipWaiting (below), or put that in a comment below?

**Review by zzmp — COMMENTED** [review]
The code looks great, can you just clean up the comments? Normally I wouldn't be this nitpicky but SW is a little esoteric, so I think it's important to document inline and link out to supporting docs (most of which you've already done 😄).

File: src/components/ErrorBoundary/index.tsx (hunk 5 above)
Comment (leggechr) [reply src/components/ErrorBoundary/index.tsx:74]: I included a link to the workbox documentation that talks about skipWaiting and why its necessary. I think this should address this comment and the one below.

**Review by zzmp — APPROVED** [review]

**Pull request merged**